--- lib/eztemplate/classes/eztemplatedofunction.php.393	2007-07-06 15:33:45.000000000 +0200
+++ lib/eztemplate/classes/eztemplatedofunction.php	2007-09-03 20:09:19.000000000 +0200
@@ -144,6 +144,11 @@
             return;
         }
 
+        include_once( "extension/ezenhanceddebug/classes/ezenhanceddebug.php" );
+        $eZEnhancedDebug = eZEnhancedDebug::instance();
+        $eZEnhancedDebugIterations = 0;
+        $eZEnhancedDebug->startTimer();
+
         do
         {
             $loop->setSequenceVar(); // set sequence variable (if specified)
@@ -154,9 +159,22 @@
                 break;
 
             $loop->incrementSequence();
+            $eZEnhancedDebugIterations ++;
         } while ( $tpl->elementValue( $functionParameters['condition'], $rootNamespace, $currentNamespace, $functionPlacement ) );
 
         $loop->cleanup();
+
+        $eZEnhancedDebug->stopTimer();
+        $eZEnhancedDebugExecutionStack = array();
+        $eZEnhancedDebugExecutionStack['operator'] = $functionName;
+        $eZEnhancedDebugExecutionStack['iterations'] = $eZEnhancedDebugIterations;
+        $eZEnhancedDebugExecutionStack['file']       = $functionPlacement[2];
+        $eZEnhancedDebugExecutionStack['linestart']  = $functionPlacement[0][0];
+        $eZEnhancedDebugExecutionStack['lineend']    = $functionChildren[ count( $functionChildren ) -1][3][1][0];
+        $eZEnhancedDebugExecutionStack['col']        = $functionPlacement[0][1];
+        $eZEnhancedDebugExecutionStack['time']       = $eZEnhancedDebug->timeDifference();
+        $eZEnhancedDebug->appendToExecutionStack( $eZEnhancedDebugExecutionStack );
+
     }
 
     /*!
--- lib/eztemplate/classes/eztemplateforeachfunction.php.393	2007-07-06 15:20:22.000000000 +0200
+++ lib/eztemplate/classes/eztemplateforeachfunction.php	2007-09-03 20:10:14.000000000 +0200
@@ -296,6 +296,10 @@
          * run the loop itself
          */
 
+        include_once( "extension/ezenhanceddebug/classes/ezenhanceddebug.php" );
+        $eZEnhancedDebug = eZEnhancedDebug::instance();
+        $eZEnhancedDebug->startTimer();
+
         /*
             $offset and $max parameters must meet the following requirements:
              -  $offset + $max <= $nItems
@@ -385,6 +389,18 @@
         }
 
         $loop->cleanup();
+
+        $eZEnhancedDebug->stopTimer();
+        $eZEnhancedDebugExecutionStack = array();
+        $eZEnhancedDebugExecutionStack['operator']   = $functionName;
+        $eZEnhancedDebugExecutionStack['iterations'] = $nItems;
+        $eZEnhancedDebugExecutionStack['file']       = $functionPlacement[2];
+        $eZEnhancedDebugExecutionStack['linestart']  = $functionPlacement[0][0];
+        $eZEnhancedDebugExecutionStack['lineend']    = $functionChildren[ count( $functionChildren ) -1][3][1][0];
+        $eZEnhancedDebugExecutionStack['col']        = $functionPlacement[0][1];
+        $eZEnhancedDebugExecutionStack['time']       = $eZEnhancedDebug->timeDifference();
+        $eZEnhancedDebug->appendToExecutionStack( $eZEnhancedDebugExecutionStack );
+
     }
 
     /*!
--- lib/eztemplate/classes/eztemplateforfunction.php.393	2007-07-06 15:59:28.000000000 +0200
+++ lib/eztemplate/classes/eztemplateforfunction.php	2007-09-03 20:11:28.000000000 +0200
@@ -197,6 +197,10 @@
             return;
         }
 
+        include_once( "extension/ezenhanceddebug/classes/ezenhanceddebug.php" );
+        $eZEnhancedDebug = eZEnhancedDebug::instance();
+        $eZEnhancedDebug->startTimer();
+
         $loop->initLoopVariable( $loopVarName );
 
         /*
@@ -221,6 +225,18 @@
         } // for
 
         $loop->cleanup();
+
+        $eZEnhancedDebug->stopTimer();
+        $eZEnhancedDebugExecutionStack = array();
+        $eZEnhancedDebugExecutionStack['operator']   = $functionName;
+        $eZEnhancedDebugExecutionStack['iterations'] = $lastVal+1;
+        $eZEnhancedDebugExecutionStack['file']       = $functionPlacement[2];
+        $eZEnhancedDebugExecutionStack['linestart']  = $functionPlacement[0][0];
+        $eZEnhancedDebugExecutionStack['lineend']    = $functionChildren[ count( $functionChildren ) -1][3][1][0];
+        $eZEnhancedDebugExecutionStack['col']        = $functionPlacement[0][1];
+        $eZEnhancedDebugExecutionStack['time']       = $eZEnhancedDebug->timeDifference();
+        $eZEnhancedDebug->appendToExecutionStack( $eZEnhancedDebugExecutionStack );
+
     }
 
     /*!
--- lib/eztemplate/classes/eztemplatewhilefunction.php.393	2007-07-06 16:09:09.000000000 +0200
+++ lib/eztemplate/classes/eztemplatewhilefunction.php	2007-09-03 20:12:13.000000000 +0200
@@ -144,6 +144,11 @@
         if ( !$loop->initialized() )
             return;
 
+        include_once( "extension/ezenhanceddebug/classes/ezenhanceddebug.php" );
+        $eZEnhancedDebug = eZEnhancedDebug::instance();
+        $eZEnhancedDebugIterations = 0;
+        $eZEnhancedDebug->startTimer();
+
         while ( $tpl->elementValue( $functionParameters['condition'], $rootNamespace, $currentNamespace, $functionPlacement ) )
         {
             $loop->setSequenceVar(); // set sequence variable (if specified)
@@ -154,9 +159,22 @@
                 break;
 
             $loop->incrementSequence();
+            $eZEnhancedDebugIterations++;
         }
 
         $loop->cleanup();
+
+        $eZEnhancedDebug->stopTimer();
+        $eZEnhancedDebugExecutionStack = array();
+        $eZEnhancedDebugExecutionStack['operator']   = $functionName;
+        $eZEnhancedDebugExecutionStack['iterations'] = $eZEnhancedDebugIterations;
+        $eZEnhancedDebugExecutionStack['file']       = $functionPlacement[2];
+        $eZEnhancedDebugExecutionStack['linestart']  = $functionPlacement[0][0];
+        $eZEnhancedDebugExecutionStack['lineend']    = $functionChildren[ count( $functionChildren ) -1][3][1][0];
+        $eZEnhancedDebugExecutionStack['col']        = $functionPlacement[0][1];
+        $eZEnhancedDebugExecutionStack['time']       = $eZEnhancedDebug->timeDifference();
+        $eZEnhancedDebug->appendToExecutionStack( $eZEnhancedDebugExecutionStack );
+
     }
 
     /*!
--- lib/eztemplate/classes/eztemplateexecuteoperator.php	2007-08-22 16:02:28.000000000 +0200
+++ lib/eztemplate/classes/eztemplateexecuteoperator.php	2007-11-06 19:46:48.000000000 +0100
@@ -355,6 +355,13 @@
             $moduleName = $namedParameters['module_name'];
             $result = eZFunctionHandler::execute( $moduleName, $functionName, $functionParameters );
             $operatorValue = $result;
+            $fetchInfos = $moduleName . '::' . $functionName;
+            $fetchRows = count( $operatorValue );
+            //var_dump( $fetchInfos, $fetchRows, $operatorValue);
+            
+            include_once( 'extension/ezenhanceddebug/classes/ezenhanceddebug.php' );
+            $eZEnhancedDebug = eZEnhancedDebug::instance();
+            $eZEnhancedDebug->FetchedRows = $fetchRows;
         }
         else if ( $operatorName == $this->FetchAlias )
         {
--- lib/ezutils/classes/ezdebug.php.40	2008-01-05 20:02:15.000000000 +0100
+++ lib/ezutils/classes/ezdebug.php	2008-01-05 20:10:25.000000000 +0100
@@ -73,6 +73,7 @@
 */
 
 //include_once( "lib/ezutils/classes/ezsys.php" );
+include_once( "extension/ezenhanceddebug/classes/ezenhanceddebug.php" );
 
 class eZDebug
 {
@@ -188,6 +189,7 @@
         $this->OverrideList = array();
         $this->topReportsList = array();
         $this->bottomReportsList = array();
+        $this->eZEnhancedDebug = eZEnhancedDebug::instance();
     }
 
     function reset()
@@ -1766,6 +1768,11 @@
             echo "</table>";
         }
 
+        if ( $as_html )
+        {
+            echo $this->eZEnhancedDebug->displayExecutionStack();
+        }
+
         $this->printBottomReportsList();
 
         if ( $as_html )
@@ -1886,6 +1893,8 @@
 
     /// A list of debug reports that appears at the top of debug output
     public $topReportsList;
+
+    private $eZEnhancedDebug;
 }
 
 /*!
--- lib/eztemplate/classes/eztemplate.php.40	2008-01-05 20:15:06.000000000 +0100
+++ lib/eztemplate/classes/eztemplate.php	2008-01-05 20:32:20.000000000 +0100
@@ -218,6 +218,7 @@
 */
 
 require_once( "lib/ezutils/classes/ezdebug.php" );
+include_once( "extension/ezenhanceddebug/classes/ezenhanceddebug.php" );
 
 //include_once( "lib/eztemplate/classes/eztemplatefileresource.php" );
 
@@ -373,6 +374,7 @@
         $this->WhileCounter   = 0;
         $this->DoCounter      = 0;
         $this->ElseifCounter  = 0;
+        $this->eZEnhancedDebug = eZEnhancedDebug::instance();
     }
 
     /*!
@@ -676,9 +678,30 @@
         {
             if ( eZTemplate::isMethodDebugEnabled() )
                 eZDebug::writeDebug( "START FUNCTION: $functionName" );
+
+            $this->eZEnhancedDebug->startTimer();
             $value = $func->process( $this, $textElements, $functionName, $functionChildren, $functionParameters, $functionPlacement, $rootNamespace, $currentNamespace );
+            $this->eZEnhancedDebug->stopTimer();
+
             if ( eZTemplate::isMethodDebugEnabled() )
                 eZDebug::writeDebug( "END FUNCTION: $functionName" );
+
+            // dirty hack
+            // excluded loop functions, will have multiple entries otherwise
+            $eZEnhancedDebugExcludedFunctions = array( 'do', 'for', 'while', 'foreach' );
+
+            if( !in_array( $functionName, $eZEnhancedDebugExcludedFunctions ) )
+            {
+                $eZEnhancedDebugExecutionStack = array();
+                $eZEnhancedDebugExecutionStack['operator']  = $functionName;
+                $eZEnhancedDebugExecutionStack['file']      = $functionPlacement[2];
+                $eZEnhancedDebugExecutionStack['linestart'] = $functionPlacement[0][0];
+                $eZEnhancedDebugExecutionStack['lineend']   = $functionPlacement[1][0];
+                $eZEnhancedDebugExecutionStack['col']       = $functionPlacement[0][1];
+                $eZEnhancedDebugExecutionStack['time']      = $this->eZEnhancedDebug->timeDifference();
+                $this->eZEnhancedDebug->appendToExecutionStack( $eZEnhancedDebugExecutionStack );
+            }
+
             return $value;
         }
         else
@@ -1376,11 +1399,24 @@
                 $value = $valueData['value'];
                 if ( eZTemplate::isMethodDebugEnabled() )
                     eZDebug::writeDebug( "START OPERATOR: $operatorName" );
+
+                $this->eZEnhancedDebug->startTimer();
                 $op->modify( $this, $operatorName, $operatorParameters, $rootNamespace, $currentNamespace, $value, $namedParameters,
                              $placement );
+                $this->eZEnhancedDebug->stopTimer();
+
                 if ( eZTemplate::isMethodDebugEnabled() )
                     eZDebug::writeDebug( "END OPERATOR: $operatorName" );
                 $valueData['value'] = $value;
+
+                $eZEnhancedDebugExecutionStack = array();
+                $eZEnhancedDebugExecutionStack['operator']  = $operatorName;
+                $eZEnhancedDebugExecutionStack['file']      = $placement[2];
+                $eZEnhancedDebugExecutionStack['linestart'] = $placement[0][0];
+                $eZEnhancedDebugExecutionStack['lineend']   = $placement[1][0];
+                $eZEnhancedDebugExecutionStack['col']       = $placement[0][1];
+                $eZEnhancedDebugExecutionStack['time']      = $this->eZEnhancedDebug->timeDifference();
+                $this->eZEnhancedDebug->appendToExecutionStack( $eZEnhancedDebugExecutionStack );
             }
             else
                 $this->error( '', "Object problem with operator '$operatorName' ",
@@ -2631,6 +2667,7 @@
     // Flag for setting compilation in test mode
     public $TestCompile;
 
+    private $eZEnhancedDebug;
 //     public $CurrentRelatedResource;
 //     public $CurrentRelatedTemplateName;
 }
